src/MonoTorrent/MonoTorrent.Client/PeerConnections/SocketConnection.cs

```diff
@@ -1,148 +1,175 @@
+
+		#endregion
+
+
+		#region Constructors
+
+		protected SocketConnection(Uri uri)
+            : this (new Socket((uri.Scheme == "ipv4") ? AddressFamily.InterNetwork : AddressFamily.InterNetworkV6, SocketType.Stream, ProtocolType.Tcp),
+                  new IPEndPoint(IPAddress.Parse(uri.Host), uri.Port), false)
+
+        {
+            Uri = uri;
+        }
+
+        protected SocketConnection(Socket socket, bool isIncoming)
+            : this(socket, (IPEndPoint)socket.RemoteEndPoint, isIncoming)
+        {
+
+        }
+
+        SocketConnection (Socket socket, IPEndPoint endpoint, bool isIncoming)
+        {
+            ReceiveArgs = new SocketAsyncEventArgs {
+                RemoteEndPoint = endpoint
+            };
+            SendArgs = new SocketAsyncEventArgs {
+                RemoteEndPoint = endpoint
+            };
+            ReceiveArgs.Completed += HandleOperationCompleted;
+            SendArgs.Completed += HandleOperationCompleted;
+            SK = socket;
+            EndPoint = endpoint;
+            IsIncoming = isIncoming;
+        }
+
+        //去掉static函数，否则导致大量的ReceiveArgs以及SendArgs不能释放--因为Dispose函数里也不能释放
+        static void HandleOperationCompleted (object sender, SocketAsyncEventArgs e)
+        {
+            if (e.SocketError != SocketError.Success)
+                ((TaskCompletionSource<int>)e.UserToken).SetException(new SocketException((int)e.SocketError));
+            else
+                ((TaskCompletionSource<int>)e.UserToken).SetResult(e.BytesTransferred);
+        }
+
+        #endregion
+
+
+        #region Async Methods
+
+        public Task ConnectAsync ()
+        {
+            var tcs = new TaskCompletionSource<int>();
+            ReceiveArgs.UserToken = tcs;
+
+            if (!SK.ConnectAsync(ReceiveArgs))
+                return Task.FromResult(true);
+            return tcs.Task;
+        }
+
+        public Task<int> ReceiveAsync(byte[] buffer, int offset, int count)
+        {
+            var tcs = new TaskCompletionSource<int>();
+            ReceiveArgs.SetBuffer(buffer, offset, count);
+            ReceiveArgs.UserToken = tcs;
+
+            if (!SK.ReceiveAsync(ReceiveArgs))
+                return Task.FromResult(ReceiveArgs.BytesTransferred);
+            return tcs.Task;
+        }
+
+        public Task<int> SendAsync(byte[] buffer, int offset, int count)
+        {
+            var tcs = new TaskCompletionSource<int>();
+            SendArgs.SetBuffer(buffer, offset, count);
+            SendArgs.UserToken = tcs;
+
+            if (!SK.SendAsync(SendArgs))
+                return Task.FromResult(SendArgs.BytesTransferred);
+            return tcs.Task;
+        }
+
+        public void Dispose()
+        {
+            try
+            {
+                if (null != ReceiveArgs)
+                {
+                    ReceiveArgs.Completed -= HandleOperationCompleted;
+                    SendArgs.Completed -= HandleOperationCompleted;

```

Comment:
If there's a pending Send or Receive then the TaskCompletionSource will never be marked as complete, which means if there's any code calling `await SendAsync` or `await ReceiveAsync` it will hang forever.

If the event handlers are left attached then any pending send/receive call will be aborted automatically, and the event handler will be invoked showing the operation has been cancelled. Similarly, if a send/receive has just completed then it will still invoke the handler and result in the TaskCompletionSource being marked as 'completed'.

I'll add a test showing what i mean...

Reply:
I popped the test and suggestions here: https://github.com/mono/monotorrent/pull/101

Reply:
I will re check the memory leaking issue,